From c0e035236696cd274f03ff15a7ca09d358aac9f1 Mon Sep 17 00:00:00 2001
From: Dmitry Kazakov <dimula73@gmail.com>
Date: Mon, 29 Jun 2020 23:41:13 +0300
Subject: [PATCH] Don't eat ShortcutOverride events when there is a
 partial-match

Some applications (e.g. Krita) may have its own shortcuts system. That
is, it should always get ShortcutOverride events, even when they are
matched to something inside Qt.

If the application (Krita) accepts the event, then all the partially
matched shortcuts should reset.

See this bug for more details:
https://bugs.kde.org/show_bug.cgi?id=409613
---
 src/gui/kernel/qshortcutmap_p.h           |  4 +++-
 src/gui/kernel/qwindowsysteminterface.cpp | 10 ++++++++--
 2 files changed, 11 insertions(+), 3 deletions(-)

diff --git a/src/gui/kernel/qshortcutmap_p.h b/src/gui/kernel/qshortcutmap_p.h
index 8fc68229..91191288 100644
--- a/src/gui/kernel/qshortcutmap_p.h
+++ b/src/gui/kernel/qshortcutmap_p.h
@@ -91,8 +91,10 @@ public:
     void dumpMap() const;
 #endif
 
-private:
     void resetState();
+
+private:
+
     QKeySequence::SequenceMatch nextState(QKeyEvent *e);
     void dispatchEvent(QKeyEvent *e);
 
diff --git a/src/gui/kernel/qwindowsysteminterface.cpp b/src/gui/kernel/qwindowsysteminterface.cpp
index b3b6167c..9bb43f2a 100644
--- a/src/gui/kernel/qwindowsysteminterface.cpp
+++ b/src/gui/kernel/qwindowsysteminterface.cpp
@@ -441,7 +441,8 @@ bool QWindowSystemInterface::handleShortcutEvent(QWindow *window, ulong timestam
         window = QGuiApplication::focusWindow();
 
     QShortcutMap &shortcutMap = QGuiApplicationPrivate::instance()->shortcutMap;
-    if (shortcutMap.state() == QKeySequence::NoMatch) {
+    if (shortcutMap.state() != QKeySequence::ExactMatch) {
+
         // Check if the shortcut is overridden by some object in the event delivery path (typically the focus object).
         // If so, we should not look up the shortcut in the shortcut map, but instead deliver the event as a regular
         // key event, so that the target that accepted the shortcut override event can handle it. Note that we only
@@ -451,8 +452,13 @@ bool QWindowSystemInterface::handleShortcutEvent(QWindow *window, ulong timestam
             QEvent::ShortcutOverride, keyCode, modifiers, nativeScanCode, nativeVirtualKey, nativeModifiers, text, autorepeat, count);
 
         {
-            if (QWindowSystemInterfacePrivate::handleWindowSystemEvent<SynchronousDelivery>(shortcutOverrideEvent))
+            if (QWindowSystemInterfacePrivate::handleWindowSystemEvent<SynchronousDelivery>(shortcutOverrideEvent)) {
+                if (shortcutMap.state() != QKeySequence::NoMatch) {
+                    shortcutMap.resetState();
+                }
+
                 return false;
+            }
         }
     }
 
-- 
2.20.1.windows.1

